kmm/kmm-post-data/src/commonMain/kotlin/com/gchristov/newsfeed/kmmpostdata/usecase/DecoratePostUseCase.kt

```diff
@@ -0,0 +1,74 @@
+package com.gchristov.newsfeed.kmmpostdata.usecase
+
+import com.gchristov.newsfeed.kmmpostdata.Post
+import com.gchristov.newsfeed.kmmpostdata.PostRepository
+import com.gchristov.newsfeed.kmmpostdata.model.DecoratedPost
+import com.gchristov.newsfeed.kmmpostdata.util.ReadingTimeCalculator
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Instant
+
+class DecoratePostUseCase(
+    private val postRepository: PostRepository,
+    private val dispatcher: CoroutineDispatcher
+) {
+
+//    suspend fun decoratedPost(
+//        postId: String
+//    ): DecoratedPost {
+//
+//        return postRepository.run {
+//            cachedPost(postId)?.let { post ->
+//                clearCache(postId)
+//                decoratePost(post)
+//            } ?: queryNewPost(postId)
+//        }
+//    }
+
+    suspend fun decoratedPost(
+        postId: String
+    ): DecoratedPost {
+
+        return postRepository.run {
+//            cachedPost(postId)?.let { post ->
+//                clearCache(postId)
+//                decoratePost(post)
+//            } ?:
+            queryNewPost(postId)
```

Comment:
I was attempting this and I think I never used these callbacks (ofc) to notify the caller, I checked
the `Feed` example, would this be in the line of what you say for `decoratedPost` in the `DecoratedPostUseCase`?

```
  suspend fun decoratedPost(
        postId: String,
        onCache: (DecoratedPost?) -> Unit): DecoratedPost {
        
        cachedPost(postId)?.let { post ->
            onCache(post)
        } ?: onCache(null)
        
        return fetchDecoratedPost(postId)
    }
    
    ...
    private suspend fun fetchDecoratedPost(postId: String): DecoratedPost =
        postRepository.run {
            post(postId).let {
                val decoratedPost = decoratePost(it)
                cachePost(decoratedPost)
                decoratedPost
            }
        }
```

Reply:
I'm figuring out how to call this from the `PostViewModel` next... :)

Reply:
Ok, how about:

```
decoratePostUseCase.apply {

                    // always show cached post
                    cachedPost(postId)?.let { post ->
                        setState { copy(post = post) }
                        clearCache(postId)
                    }

                    val newPost = decoratedPost(postId, onCache = { cachedPost ->
                        cachedPost?.let {
                            setState {
                                copy(
                                    post = cachedPost
                                )
                            }
                            // clear cache around here
                        }
                    })

                    setState {
                        copy(
                            loading = false,
                            post = newPost
                        )
                    }
                }

```

Question: why not putting the `loading` to false a this point if it's a cache hit?
The spinner is already loading while putting some pre-cached content, now again it's bringing another from the cache... can we save one of the checks? May be missing steps in the workflow still...

Reply:
Since we'll most likely going to be making only one call to the `DecoratePostUseCase` the benefits of using `apply` are not many - something like:
```
val postUseCase = ... // This is injected in the view-model as constructor parameter

val newPost = postUseCase.getPost(
    postId = postId, 
    onCache = { cachedPost ->
        cachedPost?.let {
            setState {
                copy(post = cachedPost)
            }
        }
    }
)

setState {
    copy(
        loading = false,
        post = newPost
    )
}
```
and then the definition of your `DecoratePostUseCase` could be something like
```
class DecoratePostUseCase {

    suspend fun getPost(
        postId: String,
        onCache: ((DecoratedPost) -> Unit)? // Notice that the entire callback is nullable, rather than the DecoratedPost parameter - if there's cache it will never be null. However, we might not want to care about the cache in some cases so we wouldn't want to have to pass a callback at all
): DecoratedPost {
        
        onCache?.let { cacheCallback -> // We wouldn't even run this if the caller doesn't care about cache
            cachedPost(postId)?.let { post ->
                cacheCallback(post)
            }
        }
        return fetchDecoratedPost(postId)
    }

    // This could be private so that we force the caller to either handle the cache or not in getPost, rather than having to remember to explicitly call this separate method
    private suspend fun cachedPost(postId: String): DecoratedPost? =
        postRepository.cachedPost(postId)?.let { post ->
            decoratePost(post) // Only decorate, do not clear, because we might end up in a case where we clear the local cache and the network fails. In this case the next time the same post is opened it won't be available if we have cleared it previously
        }
    
    private suspend fun fetchDecoratedPost(postId: String): DecoratedPost =
        postRepository.run {
            post(postId).let {
                val decoratedPost = decoratePost(it)
                cachePost(decoratedPost)
                decoratedPost
            }
        }
}
```

**Questions and Answers**

1. Why not putting the loading to false at the callback point if it's a cache hit?
_This is an opinionated approach of handling cache. The idea is to tell the user that something is happening while their data is refreshing. So we show the spinner initially since there's no content, then cache appears but technically we're still refreshing so we keep showing the spinner on top of the cache so they know we're not done yet. Finally when we obtain the new version of the content, we update it again and we hide the spinner. It's an approach to have better UX._

2. Why is `cachedPost` a private method in the use case?
_Again, this is an opinionated approach. The idea is to force the caller to either handler the cache or not at the call site by either passing a cache callback or not to the `getPost` method. We could do it with separate functions, but then I think it's easier to miss adding the cache logic because you have to remember to call that method yourself. Whereas if we make it a required argment on the `getPost` method we're forcing the caller to make a decision every time and explicitly pass `null` if they don't care about the cache._

Reply:
Nice one on 1) - I can see this is a pattern from UX as I went through it few times. It 'manages the expectation' you may be seeing old content, but it keep a non-blank screen

About 2) - that's what I kind of suggested, handle it all in one call (cache, non cache). In the non-usecase approach the methods to cache were exposed in the `postRepository`, so as I was refactoring it felt strange. 
Then when you suggested the `callback` approach I could see that those exposed methods (and the block of code at the top with the `apply`) 

So in the end I was reaching the same approach without idiomatic Kotlin and MVVM pattern / UX :)

Reply:
Yep exactly, looks like we're reaching the same conclusions 🙌 

I think there are different approaches for this. With coroutines and suspend functions you're sort of supposed to maybe avoid callbacks because it's not clear when they'll be called but then you'd have to put dedicated logic inside the view-model to firstly fetch the cache and then show the content. 

It's not terrible but this is sort of why you'd have a use case that encapsulates this and the callback approach just makes the call-site clearer, in my opinion, as it's all in one block (and actually follows what you'd do in similar Compose situations where it's all done with callbacks).

Reply:
This one is addressed